pkg/config/tlscfg/cert_watcher.go

```diff
@@ -195,46 +98,35 @@ func (w *certWatcher) attemptReload(rootCAs, clientCAs *x509.CertPool) {
 			)
 		}
 	}
+
+	certPairWatcher, err := fswatcher.NewFSWatcher([]string{w.opts.CertPath, w.opts.KeyPath}, onCertPairChange, w.logger)
+	if err == nil {
+		w.watchers = append(w.watchers, *certPairWatcher)
+	} else {
+		w.logger.Error(
+			"Cannot set up watcher for certificate",
+			zap.String("certificate", w.opts.CertPath),
+			zap.String("key", w.opts.KeyPath),
+			zap.Error(err),
+		)
+		w.Close()
+	}
 }
 
-func (w *certWatcher) reloadIfModified(certPath string, certHash *string, certPool *x509.CertPool) {
-	if mod, newHash := w.isModified(certPath, *certHash); mod {
+func (w *certWatcher) watchCert(certPath string, certPool *x509.CertPool) {
+	onCertChange := func() {
 		if err := addCertToPool(certPath, certPool); err == nil {
-			w.mu.Lock()
-			*certHash = newHash
-			w.mu.Unlock()
 			w.logger.Info("Loaded modified certificate", zap.String("certificate", certPath))
 		} else {
 			w.logger.Error("Failed to load certificate", zap.String("certificate", certPath), zap.Error(err))
 		}
 	}
-}
 
-// isModified returns true if the file has been modified since the last check.
-func (w *certWatcher) isModified(file string, previousHash string) (bool, string) {
-	if file == "" {
-		return false, ""
-	}
-	hash, err := hashFile(file)
-	if err != nil {
-		w.logger.Warn("Certificate has been removed, using the last known version", zap.String("certificate", file))
-		return false, ""
+	certWatcher, err := fswatcher.NewFSWatcher([]string{certPath}, onCertChange, w.logger)
+	if err == nil {
+		w.watchers = append(w.watchers, *certWatcher)
+	} else {
+		w.logger.Error("Cannot set up watcher for certificate", zap.String("certificate", certPath), zap.Error(err))
```

Comment:
this code path is executed during construction, it should propagate the error back to the caller